src/scripting/lua_kernel_base.cpp

```diff
@@ -501,6 +511,180 @@ static int intf_get_language(lua_State* L)
+			auto deprecated = luaW_toboolean(L, -1);
+			lua_pushboolean(L, deprecated);
+			return 1;
+		}
+		lua_pop(L, 1);
+	}
+	lua_pushboolean(L, false);
+	return 1;
+}
+
+// This is also a separate function so I can use a protected call on it to catch errors.
+static int impl_get_dir_suffix(lua_State*L)
+{
+	auto key = luaL_checkstring(L, 2);
+	std::string suffix = " ";
+	auto type = lua_getfield(L, 1, key);
+	if(type == LUA_TTABLE) {
+		suffix = "†";
+	} else if(type == LUA_TFUNCTION) {
+		suffix = "ƒ";
+	} else if(type == LUA_TUSERDATA) {
+		lua_getglobal(L, "getmetatable");
+		lua_pushvalue(L, -2);
+		lua_call(L, 1, 1);
+		if(lua_type(L, -1) == LUA_TSTRING) {
+			auto meta = lua_check<std::string>(L, -1);
+			if(meta == "function") {
+				suffix = "ƒ";
+			}
+		}
+	}
+	suffix = " " + suffix;
+	lua_pushlstring(L, suffix.c_str(), suffix.size());
+	return 1;
+}
+
+/**
+ * Prints out a list of keys available in an object.
+ * A list of keys is gathered from the following sources:
+ * - For a table, all keys defined in the table
+ * - Any keys accessible through the metatable chain (if __index on the metatable is a table)
+ * - The output of the __dir metafunction
+ * - Filtering out any keys beginning with two underscores
+ * - Filtering out any keys for which object[key].__deprecated exists and is true
+ * The list is then sorted alphabetically and formatted into columns.
+ * - Arg 1: Any object
+ * - Arg 2: (optional) Function to use for output; defaults to _G.print
+ */
+static int intf_object_dir(lua_State* L)
+{
+	if(lua_isnil(L, 1)) return luaL_argerror(L, 1, "Can't dir() nil");
+	if(!lua_isfunction(L, 2)) {
+		luaW_getglobal(L, "print");
+	}
+	int fcn_idx = lua_gettop(L);
+	std::vector<std::string> keys;
+	if(lua_istable(L, 1)) {
+		// Walk the metatable chain (as long as __index is a table)...
+		// If we reach an __index that's a function, check for a __dir metafunction.
+		int save_top = lua_gettop(L);
+		lua_pushvalue(L, 1);
+		ON_SCOPE_EXIT(&) {
+			lua_settop(L, save_top);
+		};
+		do {
+			int table_idx = lua_absindex(L, -1);
+			for(lua_pushnil(L); lua_next(L, table_idx); lua_pop(L, 1)) {
+				if(lua_type(L, -2) == LUA_TSTRING) {
+					keys.push_back(lua_tostring(L,-2));
+				}
+			}
+			// Two possible exit cases:
+			// 1. getmetafield returns TNIL because there is no __index
+			// In this case, the stack is unchanged, so the while condition is still true.
+			// 2. The __index is not a table
+			// In this case, obviously the while condition fails
+			if(luaL_getmetafield(L, table_idx, "__index") == LUA_TNIL) break;
+		} while(lua_istable(L, -1));
+		if(lua_isfunction(L, -1)) {
+			lua_pop(L, 1);
+			dir_meta_helper(L, keys);
+		}
+	} else if(lua_isuserdata(L, 1) && !lua_islightuserdata(L, 1)) {
+		lua_pushvalue(L, 1);
+		dir_meta_helper(L, keys);
+		lua_pop(L, 1);
+	}
+	// Sort and remove any duplicates
+	std::sort(keys.begin(), keys.end());
+	auto final = std::unique(keys.begin(), keys.end());
```

Comment:
While `final` isn't a reserved keyword in this context, I'd prefer to rename it to something that isn't a reserved word in any context. Maybe "removed"?